### PR TITLE
RBAC: Make it possible to use regex for values

### LINKED ISSUE #300: RBAC: Make it possible to use regex for values
### Issue submitter TODO list

- [X] I've searched for an already existing issues [here](https://github.com/kafbat/kafka-ui/issues)
- [X] I'm running a supported version of the application which is listed [here](https://github.com/kafbat/kafka-ui/blob/main/.github/SECURITY.md) and the feature is not present there

### Is your proposal related to a problem?

_No response_

### Describe the feature you're interested in

RBAC to support regex values, so you could use "*@domain.tld" emails to match the roles without having to populate the users with roles 

### Describe alternatives you've considered

_No response_

### Version you're running

27c5e61

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by francoisvandenplas**
- [ ] **Breaking change?** (if so, please describe the impact and migration path for existing application instances)

**What changes did you make?** (Give an overview)
Rbac now handle regex in values
Fixes #300

**Is there anything you'd like reviewers to focus on?**

**How Has This Been Tested?** (put an "x" (case-sensitive!) next to an item)

- [ ] No need to
- [ ] Manually (please, describe, if necessary)
- [x] Unit checks
- [ ] Integration checks
- [ ] Covered by existing automation

**Checklist** (put an "x" (case-sensitive!) next to all the items, otherwise the build will fail)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (e.g. **ENVIRONMENT VARIABLES**)
- [x] My changes generate no new warnings (e.g. Sonar is happy)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged

Check out [Contributing](https://github.com/kafbat/kafka-ui/blob/main/.github/CONTRIBUTING.md) and [Code of Conduct](https://github.com/kafbat/kafka-ui/blob/main/.github/CODE-OF-CONDUCT.md)

**A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by callaertanthony** [comment]
Hey ! Any chance to see this PR merged ?

**Comment by Haarolean** [comment]
> Hey ! Any chance to see this PR merged ?

sorry, we'll take a look this week

File: api/src/main/java/io/kafbat/ui/service/rbac/extractor/CognitoAuthorityExtractor.java [hunk 1]
```diff
@@ -50,7 +50,7 @@ private Set<String> extractUsernameRoles(AccessControlService acs, DefaultOAuth2
             .stream()
             .filter(s -> s.getProvider().equals(Provider.OAUTH_COGNITO))
             .filter(s -> s.getType().equals("user"))
-            .anyMatch(s -> s.getValue().equalsIgnoreCase(principal.getName())))
+            .anyMatch(s -> principal.getName() != null && principal.getName().matches(s.getValue())))
```
Comment (Haarolean) [inline api/src/main/java/io/kafbat/ui/service/rbac/extractor/CognitoAuthorityExtractor.java:53]: this can't be null, there should be a condition check in `DefaultOAuth2User`

File: api/src/test/java/io/kafbat/ui/config/RegexBasedProviderAuthorityExtractorTest.java [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    provider.setCustomParams(Map.of("roles-field", "role_definition"));
+    additionalParams.put("provider", provider);
+
+    Set<String> roles = extractor.extract(accessControlService, oauth2User, additionalParams).block();
+
+    assertEquals(Set.of("viewer", "admin"), roles);
+
+  }
+
+}
```
Comment (Haarolean) [inline api/src/test/java/io/kafbat/ui/config/RegexBasedProviderAuthorityExtractorTest.java:159]: can we add a few tests to check negative scenarios (like user doesn't get assigned roles he shouldn't have; nothing explodes with unexpected values)?

**Review by Haarolean — CHANGES_REQUESTED** [review]
Thanks for covering this with tests, that's really helpful!
Left a few minor inline comments

File: api/src/test/java/io/kafbat/ui/config/RegexBasedProviderAuthorityExtractorTest.java (hunk 2 above)
Comment (francoisvandenplas) [reply api/src/test/java/io/kafbat/ui/config/RegexBasedProviderAuthorityExtractorTest.java:194]: I added some tests and assertions.
We should get _PatternSyntaxException_ if pattern has unexpected values.

File: api/src/main/java/io/kafbat/ui/service/rbac/extractor/OauthAuthorityExtractor.java [hunk 3]
```diff
@@ -59,8 +59,8 @@ private Set<String> extractUsernameRoles(AccessControlService acs, DefaultOAuth2
             .filter(s -> s.getProvider().equals(Provider.OAUTH))
             .filter(s -> s.getType().equals("user"))
             .peek(s -> log.trace("[{}] matches [{}]? [{}]", s.getValue(), principalName,
-                s.getValue().equalsIgnoreCase(principalName)))
-            .anyMatch(s -> s.getValue().equalsIgnoreCase(principalName)))
+                principalName != null && principalName.matches(s.getValue())))
+            .anyMatch(s -> principalName != null && principalName.matches(s.getValue())))
```
Comment (Haarolean) [inline api/src/main/java/io/kafbat/ui/service/rbac/extractor/OauthAuthorityExtractor.java:63]: same here

**Review by Haarolean — CHANGES_REQUESTED** [review]

**Comment by callaertanthony** [comment]
hey @Haarolean , is it possible for you to give a look again ? many thanks

**Comment by Haarolean** [comment]
@callaertanthony thank you! If you have time, could you update our docs as well? https://github.com/kafbat/ui-docs

**Comment by Haarolean** [comment]
@callaertanthony hey, good point! We've discussed this: we believe a better approach would be introducing a new boolean param like `isRegex` (nullable & defaults to false for back compatibility), which will indicate whether we should treat the value as regex or a string literal. This will guarantee backward compatibility AND reduce confusion if multiple fields are used instead. What do you think?

**Comment by callaertanthony** [comment]
thanks @Haarolean , @francoisvandenplas I created this pr with the change : https://github.com/francoisvandenplas/kafka-ui/pull/1

**Comment by Haarolean** [comment]
> thanks @Haarolean , @francoisvandenplas I created this pr with the change : [francoisvandenplas#1](https://github.com/francoisvandenplas/kafka-ui/pull/1)

could you rebase this branch instead please?

**Review by Haarolean — CHANGES_REQUESTED** [review]

File: api/src/main/java/io/kafbat/ui/model/rbac/Subject.java [hunk 4]
```diff
@@ -4,32 +4,37 @@
...
 
 @Getter
 public class Subject {
 
   Provider provider;
+  @Setter
   String type;
+  @Setter
   String value;
+  @Setter
```
[the remark this answers is not in the document]
Comment (francoisvandenplas) [reply api/src/main/java/io/kafbat/ui/model/rbac/Subject.java]: I replace with `@Data` as _Yaml.load()_ needs an empty constructor in RegexBasedProviderAuthorityExtractorTest

[the remark this answers is not in the document]
Comment (Haarolean) [reply api/src/main/java/io/kafbat/ui/model/rbac/Subject.java]: @francoisvandenplas can we use jackson instead? it can work with parametrized constructors.

[the remark this answers is not in the document]
Comment (francoisvandenplas) [reply api/src/main/java/io/kafbat/ui/model/rbac/Subject.java]: Yes, we could, but we have to provide a constructor with _@jsonCreator_ annotation as :
```
  @JsonCreator
  public Subject(
      @JsonProperty("provider") Provider provider,
      @JsonProperty("type") String type,
      @JsonProperty("value") String value,
      @JsonProperty("isRegex") boolean isRegex
  ) {
    this.provider = provider;
    this.type = type;
    this.value = value;
    this.isRegex = isRegex;
  } 
```

[the remark this answers is not in the document]
Comment (francoisvandenplas) [reply api/src/main/java/io/kafbat/ui/model/rbac/Subject.java]: There is some test failures.
rbac classes are also mapped by config properties hence the` @Data` on those classes.
Any idea ?

[the remark this answers is not in the document]
Comment (callaertanthony) [reply api/src/main/java/io/kafbat/ui/model/rbac/Subject.java]: hello @francoisvandenplas , sorry, I don't have enough time to go further, but rollback to getter annotation and specific setters works : https://github.com/francoisvandenplas/kafka-ui/pull/2

[the remark this answers is not in the document]
Comment (Haarolean) [reply api/src/main/java/io/kafbat/ui/model/rbac/Subject.java]: `@Data` itself generates setters for non-final fields and a constructor. So rolling back to using only `@Data` and loading a test file with jackson rather than snakeyaml should help.

**Review by Haarolean — CHANGES_REQUESTED** [review]
just one minor thing and we're good to merge. @francoisvandenplas would you mind updating docs as well? 
https://github.com/kafbat/ui-docs

**Review by germanosin — APPROVED** [review]

File: api/src/main/java/io/kafbat/ui/model/rbac/Subject.java [hunk 5]
```diff
@@ -3,27 +3,19 @@
...
-  }
-
-  public void setType(String type) {
-    this.type = type;
-  }
-
-  public void setValue(String value) {
-    this.value = value;
-  }
+  @JsonProperty("isRegex")
```
[the remark this answers is not in the document]
Comment (Haarolean) [reply api/src/main/java/io/kafbat/ui/model/rbac/Subject.java:17]: Ah, that seems invalid, as jackson will add an extra `is` prefix

**Comment by Haarolean** [comment]
@francoisvandenplas @callaertanthony thanks for your contribution! We just need to update docs now
